### DISCUSSION TITLE
How do we create Trait Offers with the SDK?

### DISCUSSION THREAD

**Discussion by Emanuele-Fortunati** [ProjectOpenSea/opensea-js#1410] [Q&A]
I've been trying for a while now, but I'm unable to create a Trait Offers on a collection with the SDK. Any idea on how to do it?

**Comment by ryanio** [comment]
This is something we still have to add to the SDK, for now you can use https://docs.opensea.io/reference/post_criteria_offer_v2 and the internal "get" helper in the sdk which will add the api key for you.

**Reply by Emanuele-Fortunati** [reply]
Hello! Thanks for your answer. I’ve actually did try to use that endpoint directly, however there are some parameters I can’t figure out how to set. And if I test on the api page, the errors are incorrect in some cases, so they won’t help me.
Could you post an example on how to fill the parameters?
Most of them I guess come from the build offer endpoint (which I managed to use properly), but the ones that are not there, I can’t really set them.
Thanks

**Reply by ryanio** [reply]
what parameters are you confused by? only the ones with * are required

**Reply by Emanuele-Fortunati** [reply]
@ryanio  Hey Ryan, thanks for your support. I'll show you an example, I think it's easier.
I'm trying to make an offer on Nakamigos, with criteria "Type = Frog"

So, first of all I make the call to build the offer, like:

`const options = {
  method: 'POST',
  headers: {
    accept: 'application/json',
    'content-type': 'application/json',
    'x-api-key': '%MY_API_KEY%'
  },
  body: JSON.stringify({
    quantity: 1,
    criteria: {collection: {slug: 'nakamigos'}, trait: {type: 'Type', value: 'Frog'}},
    offer_protection_enabled: true,
    offerer: '%MY_WALLET_ADDRESS%',
    protocol_address: '0x00000000000000adc04c56bf30ac9d3c0aaf14dc'
  })
};

const offerData = await fetch('https://api.opensea.io/api/v2/offers/build', options);
const offer = await offerData.json();`
and I get the following response:
`{
  "partialParameters": {
    "consideration": [
      {
        "itemType": %ITEM_TYPE%,
        "token": "%TOKEN%",
        "identifierOrCriteria": "%IDENTIFIER_OR_CRITERIA%",
        "startAmount": "1",
        "endAmount": "1",
        "recipient": "%%RECIPIENT%"
      }
    ],
    "zone": "%ZONE%",
    "zoneHash": "%ZONE_HASH%"
  },
  "encoded_token_ids": "%ENCODED_TOKEN_IDS%"
}`
(By the way, if I use openseaSDK.api.post('/api/v2/offers/build', offerParams) it returns an error, but the error object is empty)

Now, to make the offer, I use the following body:
`{
  protocol_data: {
      parameters: {
        orderType: 0,
        offerer: '%MY_WALLET_ADDRESS%',
        offer: [
          {
            itemType: %ITEM_TYPE%,
            identifierOrCriteria: '%IDENTIFIER_OR_CRITERIA%',
            startAmount: 0.0013,
            endAmount: 0.0013,
            token: '%TOKEN%'
          }
        ],
        startTime: 1710228960,
        endTime: 1710228980,
        zone: '%ZONE%',
        zoneHash: '%ZONE_HASH%',
        salt: '%SALT%',
        conduitKey: '0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000',
        totalOriginalConsiderationItems: ????,
        counter: '??????',
        consideration: [
          {
            itemType: %ITEM_TYPE%,
            token: '%TOKEN%',
            identifierOrCriteria: '%IDENTIFIER_OR_CRITERIA%',
            startAmount: 0.0013,
            endAmount: 0.0013,
            recipient: '????????'
          }
        ]
      },
      signature: '???????'
    },
    criteria: {collection: {slug: 'nakamigos'}, trait: {type: 'Type', value: 'Frog'}},
    protocol_address: '0x00000000000000adc04c56bf30ac9d3c0aaf14dc'
}`

So, I'm unsure about:
- counter
- totalOriginalConsiderationItems (it may be %ENCODED_TOKEN_IDS%.legth ??) 
- recipient
- signature

But also, if I use itemType: %ITEM_TYPE%,, with %ITEM_TYPE% = 4, which is what I got from the build endpoint, I get the following error
        "Only ERC20 tokens are supported for making offers, got 4 and expected 1"
If I put 1, I get the following error
        "identifierOrCriteria for offer item must be 0, got %IDENTIFIER_OR_CRITERIA%"
If I put identifierOrCriteria = 0, I get another error.... and so on.

So it is a bit complicated really and the errors eventually end up in a loop, that's why I've asked for help here.

Any help would be great, thanks!

**Comment by Emanuele-Fortunati** [comment]
@ryanio  I've made some progresses by checking out the code you guys created for the SDK.
Now the last bit of error I get is "You provided an invalid item type: 4, please use a non-criteria item type" - however I know we can now create criteria (trait) offers.
Any idea on how to solve this last bit?

Here the offer object:

`{
  offer: [
    {
      token: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      amount: '1000000000000000',
      endAmount: '1000000000000000'
    }
  ],
  consideration: [
    {
      itemType: 4,
      token: '0xd774557b647330C91Bf44cfEAB205095f7E6c367',
      identifierOrCriteria: '40493282301148232205507986575643870319285550606441004248024940820055161532964',
      amount: '1',
      endAmount: '1',
      recipient: '0x9c0f056729a31F323808d9a61BD7246170c1185d'
    },
    {
      token: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
      amount: '25000000000000',
      endAmount: '25000000000000',
      recipient: '0x0000a26b00c1f0df003000390027140000faa719'
    }
  ],
  endTime: '1711394158',
  zone: '0x000000e7ec00e7b300774b00001314b8610022b8',
  zoneHash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  salt: '0',
  restrictedByZone: true,
  allowPartialFills: true
}`

Thanks.

**Comment by ryanio** [comment]
Trait offers are now supported in opensea-js, please update to the newest version to use!

**Reply by Emanuele-Fortunati** [reply]
Amzing!
